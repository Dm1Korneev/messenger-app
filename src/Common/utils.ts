
type Filter = Parameters<typeof Array.prototype.filter>[0]
export const distinct: Filter = (value, index, self) => self.indexOf(value) === index;
