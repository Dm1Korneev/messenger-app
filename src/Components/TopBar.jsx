import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import withStyles from '@material-ui/core/styles/withStyles';
import Typography from '@material-ui/core/Typography';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';

import MenuIcon from '@material-ui/icons/Menu';
import CreateIcon from '@material-ui/icons/Create';

import { DRAWER_WIDTH } from 'Constants';

function TopBar(props) {
  const {
    drawerIsOpen,
    user,
    onLogout,
    onDriwerOpen,
    openModifyUserDialog,
    classes,
  } = props;
  const userName = user && user.name;

  return (
    <AppBar
      position="absolute"
      className={classNames(
        classes.appBar,
        drawerIsOpen && classes.appBarShift,
      )}
    >
      <Toolbar disableGutters={!drawerIsOpen} className={classes.toolbar}>
        <IconButton
          color="inherit"
          aria-label="Open drawer"
          onClick={onDriwerOpen}
          className={classNames(
            classes.menuButton,
            drawerIsOpen && classes.menuButtonHidden,
          )}
        >
          <MenuIcon />
        </IconButton>
        <Typography
          component="h1"
          variant="h6"
          color="inherit"
          noWrap
          className={classes.title}
        >
          {userName}
          <IconButton color="inherit" onClick={openModifyUserDialog}>
            <CreateIcon />
          </IconButton>
        </Typography>
        <Button color="inherit" onClick={onLogout}>
          logout
        </Button>
      </Toolbar>
    </AppBar>
  );
}

TopBar.propTypes = {
  classes: PropTypes.instanceOf(Object).isRequired,
  drawerIsOpen: PropTypes.bool.isRequired,
  onLogout: PropTypes.func.isRequired,
  onDriwerOpen: PropTypes.func.isRequired,
  openModifyUserDialog: PropTypes.func.isRequired,
  user: PropTypes.shape({
    name: PropTypes.string,
  }).isRequired,
};

const styles = (theme) => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    boxShadow: 'none',
    borderBottom: '1px solid '.concat(theme.palette.divider),
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  appBarShift: {
    marginLeft: DRAWER_WIDTH,
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
  toolbar: {
    paddingRight: 24, // keep right padding when drawer closed
  },
  menuButton: {
    marginLeft: 12,
    marginRight: 36,
  },
  menuButtonHidden: {
    display: 'none',
  },
  title: {
    flexGrow: 1,
  },
});

export default withStyles(styles)(TopBar);
