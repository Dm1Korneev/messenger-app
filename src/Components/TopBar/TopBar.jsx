import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import Box from '@material-ui/core/Box';
import MenuIcon from '@material-ui/icons/Menu';
import CreateIcon from '@material-ui/icons/Create';

import { DRAWER_WIDTH } from 'Constants';

const useStyles = makeStyles((theme) => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  appBarShift: {
    marginLeft: DRAWER_WIDTH,
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
}));

function TopBar(props) {
  const classes = useStyles();

  const {
    drawerIsOpen,
    user,
    onLogout,
    onDriwerOpen,
    openModifyUserDialog,
  } = props;
  const userName = user && user.name;

  return (
    <AppBar
      position="fixed"
      className={classNames(
        classes.appBar,
        drawerIsOpen && classes.appBarShift,
      )}
    >
      <Toolbar>
        <Box mr={2} hidden={drawerIsOpen}>
          <IconButton
            color="inherit"
            edge="start"
            aria-label="Open drawer"
            onClick={onDriwerOpen}
          >
            <MenuIcon />
          </IconButton>
        </Box>
        <Box flexGrow="1">
          <Typography
            variant="h6"
            color="inherit"
          >
            {userName}
            <IconButton color="inherit" onClick={openModifyUserDialog}>
              <CreateIcon />
            </IconButton>
          </Typography>
        </Box>
        <Button color="inherit" onClick={onLogout}>
          logout
        </Button>
      </Toolbar>
    </AppBar>
  );
}

TopBar.propTypes = {
  drawerIsOpen: PropTypes.bool.isRequired,
  onLogout: PropTypes.func.isRequired,
  onDriwerOpen: PropTypes.func.isRequired,
  openModifyUserDialog: PropTypes.func.isRequired,
  user: PropTypes.shape({
    name: PropTypes.string,
  }).isRequired,
};

export default TopBar;

