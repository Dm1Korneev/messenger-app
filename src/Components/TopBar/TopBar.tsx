import React from 'react';
import clsx from 'clsx';
import { makeStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Button from '@material-ui/core/Button';
import Box from '@material-ui/core/Box';
import MenuIcon from '@material-ui/icons/Menu';
import CreateIcon from '@material-ui/icons/Create';
import { useDispatch, useSelector } from 'react-redux';

import * as Actions from 'Redux/actions';
import { currentUserSelector, drawerIsOpenSelector } from 'Selectors/session';
import { DRAWER_WIDTH } from 'Constants';

const useStyles = makeStyles((theme) => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
  },
  appBarShift: {
    marginLeft: DRAWER_WIDTH,
    width: `calc(100% - ${DRAWER_WIDTH}px)`,
    transition: theme.transitions.create(['width', 'margin'], {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.enteringScreen,
    }),
  },
}));

export const TopBar = () => {
  const dispatch = useDispatch();

  const currentUser = useSelector(currentUserSelector);
  const drawerIsOpen = useSelector(drawerIsOpenSelector);

  const classes = useStyles();

  const onDrawerOpen = () => dispatch(Actions.setDrawerIsOpen(true));
  const onLogout = () => dispatch(Actions.logOut());
  const openModifyUserDialog = () => dispatch(Actions.setModifyUserDialogIsOpen(true));

  const userName = currentUser?.name;

  return (
    <AppBar
      position="fixed"
      className={clsx(
        classes.appBar,
        {
          [classes.appBarShift]: drawerIsOpen,
        },
      )}
    >
      <Toolbar>
        <Box mr={2} hidden={drawerIsOpen}>
          <IconButton
            color="inherit"
            edge="start"
            aria-label="Open drawer"
            onClick={onDrawerOpen}
          >
            <MenuIcon />
          </IconButton>
        </Box>
        <Box flexGrow="1">
          <Typography
            variant="h6"
            color="inherit"
          >
            {userName}
            <IconButton color="inherit" onClick={openModifyUserDialog}>
              <CreateIcon />
            </IconButton>
          </Typography>
        </Box>
        <Button color="inherit" onClick={onLogout}>
          logout
        </Button>
      </Toolbar>
    </AppBar>
  );
};
