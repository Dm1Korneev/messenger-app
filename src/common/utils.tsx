export function distinct(value: any, index: number, self: Array<any>): boolean {
  return self.indexOf(value) === index;
}
