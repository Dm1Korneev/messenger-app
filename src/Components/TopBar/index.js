import commonHoc from 'Components/commonHoc';

import {
  logOut,
  setDrawerIsOpen,
  setModifyUserDialogIsOpen,
} from 'Redux/actions';

import { currentUserSelector, drawerIsOpenSelector } from 'Selectors/session';

import TopBar from './TopBar';

const mapStateToProps = (state) => ({
  user: currentUserSelector(state),
  drawerIsOpen: drawerIsOpenSelector(state),
});

const mapDispatchToProps = {
  onDriwerOpen: () => setDrawerIsOpen(true),
  onLogout: logOut,
  openModifyUserDialog: () => setModifyUserDialogIsOpen(true),
};

const Container = commonHoc(TopBar, {
  mapStateToProps,
  mapDispatchToProps,
});

export { Container as TopBar };
