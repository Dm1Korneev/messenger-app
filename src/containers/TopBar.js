import { connect } from 'react-redux';
import {
  logOut,
  setDrawerIsOpen,
  setUserModifyDialogIsOpen,
} from '../redux/actions';
import TopBar from '../components/TopBar';

const mapStateToProps = (state) => ({
  user: state.session.user,
  drawerIsOpen: state.session.drawerIsOpen,
});

const mapDispatchToProps = (dispatch) => ({
  onDriwerOpen: () => dispatch(setDrawerIsOpen(true)),
  onLogout: () => dispatch(logOut()),
  openUserModifyDialog: () => dispatch(setUserModifyDialogIsOpen(true)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TopBar);
