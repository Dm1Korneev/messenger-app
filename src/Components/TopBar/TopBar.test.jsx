import React from 'react';
import { render } from '@testing-library/react';
import { mockComponent } from 'testing/utils';

import TopBar from './TopBar';

jest.mock('@material-ui/core/Typography', () => mockComponent('Typography'));
jest.mock('@material-ui/core/AppBar', () => mockComponent('AppBar'));
jest.mock('@material-ui/core/Toolbar', () => mockComponent('Toolbar'));
jest.mock('@material-ui/core/IconButton', () => mockComponent('IconButton'));
jest.mock('@material-ui/core/Button', () => mockComponent('Button'));
jest.mock('@material-ui/icons/Menu', () => mockComponent('Menu'));
jest.mock('@material-ui/icons/Create', () => mockComponent('Create'));

const props = {
  drawerIsOpen: false,
  onLogout: jest.fn(),
  onDriwerOpen: jest.fn(),
  openModifyUserDialog: jest.fn(),
  user: {
    name: 'name',
  },
};

describe('render TopBar component', () => {
  const setup = () => render(
    <TopBar {...props} />,
  );

  test('snapshot test', async () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });
});
