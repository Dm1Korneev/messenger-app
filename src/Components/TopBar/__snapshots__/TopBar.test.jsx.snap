// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render TopBar component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="AppBar"
    test-prop-classname="\\"TopBar-appBar-1\\""
    test-prop-position="\\"absolute\\""
  >
    <div
      data-originalcomponent="Toolbar"
      test-prop-classname="\\"TopBar-toolbar-3\\""
      test-prop-disablegutters="true"
    >
      <div
        data-originalcomponent="IconButton"
        test-prop-aria-label="\\"Open drawer\\""
        test-prop-classname="\\"TopBar-menuButton-4\\""
        test-prop-color="\\"inherit\\""
      >
        <div
          data-originalcomponent="Menu"
        />
      </div>
      <div
        data-originalcomponent="Typography"
        test-prop-classname="\\"TopBar-title-6\\""
        test-prop-color="\\"inherit\\""
        test-prop-component="\\"h1\\""
        test-prop-nowrap="true"
        test-prop-variant="\\"h6\\""
      >
        name
        <div
          data-originalcomponent="IconButton"
          test-prop-color="\\"inherit\\""
        >
          <div
            data-originalcomponent="Create"
          />
        </div>
      </div>
      <div
        data-originalcomponent="Button"
        test-prop-color="\\"inherit\\""
      >
        logout
      </div>
    </div>
  </div>
</DocumentFragment>
`;
